keps/sig-auth/4317-pod-certificates/README.md

```diff
@@ -0,0 +1,1313 @@
+CertificateSigningRequests.  There are no separate approval and signing stages.
+They can only be issued (and thus implicitly approved), denied (represented by a "Denied" status), or failed (represented by a "Failed" status).
+
+Similar to CertificateSigningRequests, controllers that want to issue a
+certificate in response to a PodCertificateRequest need to have two separate RBAC permissions:
+* `update` on `podcertificaterequests/status`.
+* `sign` on the requested signer.
+
+Once a signer controller decides that a PodCertificateRequest is addressed to it by matching on the signer name, it should take one of the following actions:
+* Issue the certificate by updating the status subresource of the
+  PodCertificateRequest, setting both the `certificateChain` and
+  `beginRefreshAt` fields.
+* Reject the certificate by updating the status subresource of the
+  PodCertificateRequest, adding a "Denied" condition.
+* Indicate failure in issuing the certificate by updating the status subresource
+  of the PodCertificateRequest, adding a "Failed" condition.
+
+PodCertificateRequest validation logic will:
+* Confirm that the public key is one of the supported key types.
+* Confirm that the proof-of-possession is valid.
+* Confirm that the issued chain (if one is set) consists of valid certificates,
+  that do indeed form a chain to each other.
+* Confirm that neither or both of the certificateChain and beginRefreshAt
+  timestamp are set.
+* Ensure that all fields in PodCertificateRequest.Spec are immutable.
+* Ensure that all fields in PodCertificateRequest.Status are immutable once the
+  request has been issued, denied, or failed.
+
+The noderestriction admission plugin checks the consistency of the information in created PodCertificateRequest objects:
+* The namespace, pod name/UID, service account name/UID, and node name/UID must
+  correspond to a real pod that currently exists in the cluster.
+* The pod should be in the Pending or Running phases.
+* The PodCertificateRequest must have been created by the user
+  `system:node:foo`, where `foo` is the node named in the PodCertificateRequest
+  spec.
+
+This upholds the Kubernetes node isolation guarantee by ensuring that
+compromising a single node only results in compromises of workloads that have
+actually been scheduled onto that node.
+
+During alpha, the ability to store PodCertificate objects (and the node
+restriction admission logic) will be gated by the PodCertificateRequests feature
+gate.
+
+#### Why a New Type?
+
+The following constraints led to the adoption of a new API type, as opposed to extending CertificateSigningRequest:
+1. We wanted the pod details to be legible without having to parse an ASN.1 CSR.
+2. We wanted the pod details (especially the associated node name) to be
+  available for use in an informer selector.  While these could be surfaced as labels on CertificateSigningRequest objects, labels are immutable.
```

Comment:
Do you mean the labels are mutable here?

Reply:
This holds.

Reply:
Yes, I meant to say labels are *mutable*.  Updating.

Reply:
Done.